components/eamxx/src/control/atmosphere_driver.cpp

```diff
@@ -985,291 +985,109 @@ void AtmosphereDriver::set_initial_conditions ()
+      // lat/lon column to every other column
+      auto fnames = model_init->get_ic_fields_in_file_names ();
+      auto fm = get_field_mgr(model_init->get_ic_grid()->name());
+      iop->setup_io_info(filename, model_init->get_ic_grid());
+      iop->read_fields_from_file_for_iop(filename, fnames, m_current_ts, fm);
+    } else {
+      model_init->read_ic_file (filename,m_atm_logger);
     }
+    m_atm_logger->info("    [EAMxx] Reading initial conditions file ... done!");
   }
-  m_atm_logger->debug("    [EAMxx] Processing subfields ... done!");
 
-  // Load topography from file if topography file is given.
   if (ic_pl.isParameter("topography_filename")) {
-    m_atm_logger->info("    [EAMxx] Reading topography from file ...");
-    const auto& file_name = ic_pl.get<std::string>("topography_filename");
-    m_atm_logger->info("        filename: " + file_name);
-    for (const auto& it : m_field_mgrs) {
-      const auto& grid_name = it.first;
-      if (not m_intensive_observation_period) {
-        // Topography files always use "ncol_d" for the GLL grid value of ncol.
-        // To ensure we read in the correct value, we must change the name for that dimension
-        auto io_grid = it.second->get_grid();
-        if (grid_name=="Physics GLL") {
-          using namespace ShortFieldTagsNames;
-          auto grid = io_grid->clone(io_grid->name(),true);
-          grid->reset_field_tag_name(COL,"ncol_d");
-          io_grid = grid;
-        }
-        read_fields_from_file (topography_file_fields_names[grid_name],
-                               topography_eamxx_fields_names[grid_name],
-                               io_grid,file_name,m_current_ts);
-      } else {
-        // For IOP enabled, we load from file and copy data from the closest
-        // lat/lon column to every other column
-        m_intensive_observation_period->read_fields_from_file_for_iop(file_name,
-                                                                      topography_file_fields_names[grid_name],
-                                                                      topography_eamxx_fields_names[grid_name],
-                                                                      m_current_ts,
-                                                                      it.second);
+    const auto& filename = ic_pl.get<std::string>("topography_filename");
+    m_atm_logger->info("    [EAMxx] Reading topography file ...");
+    m_atm_logger->info("        filename: " + filename);
+    m_atm_params.sublist("provenance").set("topography_file",filename);
+    if (iop) {
+      // For IOP, we load from file and copy data from the closest
+      // lat/lon column to every other column
+      auto topo_fnames_file  = model_init->get_topo_fields_names_file();
+      auto topo_fnames_eamxx = model_init->get_topo_fields_names_eamxx();
+      for (const auto& it : topo_fnames_file) {
+        const auto& gname = it.first;
+        auto fm = get_field_mgr(gname);
```

Comment:
I think here we will need to insert
```
if (gname == "Physics PG2") {
  iop->setup_io_info(filename, fm->get_grid());
}
```

Reply:
Looks like it's safe to call `setup_io_info` multiple times (the if statements should prevent code from running twice), so maybe I don't need to check the grid name? (I'm trying to keep explicit knowledge about grids names as much as possible outside the driver...)

Reply:
The issue is that if `setup_io_info()` is called on GLL grid with the topo file it should error out, because topo file has `lat`/`lon` for PG2 only. Maybe if GLL is already setup (it will be because ICs are read first) we will just return, but that kinda hides that it is an incompatible call.

Another option would be to change `setup_io_info()` to take in a list of grids that require read from file, and the IC file and topo file (if exists), and then call before ICs/topo are read. Then internally IOP can error out if an incorrect request is made (e.g., read from PG2 but no topo file) and setup every grid needed at once.

Reply:
And from the AD's perspective, they just hand over all the info they have agnostic of exactly which grids they are.

Reply:
Ok, I will think of how to best handle this.